### PR TITLE
Python: rework documentation

### PR DESCRIPTION
**Pull request by yoff**
This PR is a step towards the structure suggested in https://github.com/github/codeql-python-team/issues/319. It

- rearranges sections
- removes the unused sections
- adds a new section on data flow
- removes all references to the old data-flow implementation from surrounding documents

Remaining sections can then be added later. The section on data flow is an adaptation from the one for C# with exercises removed. It has running examples and should be accurate if not comprehensive. For example, it only introduces `LocalSourceNode` very briefly (but does illustrate the concept with an example).

I am not sure if we are ready to merge this PR as is or if we would rather build a tower of PRs on top, adding the remaining planned sections. But I open it up for review now, so people (especially from the doc team) can get involved.

### CODE REVIEW

**Comment by yoff** [comment]
Does this need a change-note? It is not really tied to the code, right?

File: .vscode/settings.json
Comment (shati-patel) [inline .vscode/settings.json:3]:
```suggestion
    "omnisharp.autoStart": false
```
Is it okay to revert this rST config change? It might cause difficulties if people have a different workspace root 🙂

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:177]:
```suggestion
   import python
```

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:215]:
```suggestion
   import python
```

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:250]:
```suggestion
   import python
```

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:251]:
```suggestion
   import semmle.code.python.dataflow.flowsources.PublicCallableParameter
```

I'm assuming the rest of the import path is correct? Or should it be `semmle.python.dataflow.new.RemoteFlowSources`?

File: docs/codeql/codeql-language-guides/codeql-for-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/codeql-for-python.rst:22]:
```suggestion
-  :doc:`Analyzing data flow in Python <analyzing-data-flow-in-python>`: You can use CodeQL to track the flow of data through a Python program to places where the data is used.
```

(Copied from the new page's intro)

**Review by shati-patel — COMMENTED** [review]
Thanks for this update! A couple of small suggestions, but otherwise LGTM. It's great to see more consistency across the languages 😃

> Does this need a change-note? It is not really tied to the code, right?

You're right, the changes you've made here don't need a change note! I'm assuming there was a change note when you added the new libraries themselves?

> I am not sure if we are ready to merge this PR as is or if we would rather build a tower of PRs on top, adding the remaining planned sections. But I open it up for review now, so people (especially from the doc team) can get involved.

From my perspective, it's fine to merge as-is. Assuming the information/examples are accurate (I defer to the Python team for that 😅), this looks like a good change in itself. We can always expand on the details and add more sections later!

File: .vscode/settings.json
Comment (yoff) [reply .vscode/settings.json:3]: Oops, did mean to commit that!

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:251]: Yes, it should be the latter, thank you for catching all these non-runnable examples!

**Review by shati-patel — APPROVED** [review]
Looks good!

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (tausbn) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:122]: This is a bit clunky. Would it perhaps be better to simply encourage people to use `hasLocalSource` instead?

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (tausbn) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:140]: Hot take: we should be able to use `flowsTo` here, and hence `ParameterNode` should in fact extend `LocalSourceNode`.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (tausbn) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:156]: In light of my comment in the data-flow version of this example, I wonder if we should have `taintFlowsTo` as a method on `LocalSourceNode` and perhaps discourage the use of `localFlow`/`localTaint` entirely.

**Review by tausbn — CHANGES_REQUESTED** [review]
Generally this looks good to me, but I feel like we should be emphasising `LocalSourceNode` more as the right abstraction to use. If we can get people started on using these early, then type trackers can be presented as using `LocalSourceNode` everywhere (and hence reap the corresponding performance benefits).

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:122]: That could look a fair bit nicer, then we would have to expose it, though.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:140]: That would make a lot of sense, morally all parameters are local sources. I am not sure if some technical SSA/CfgNode discrepancy has to be overcome first..

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:156]: That might be good.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (tausbn) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:122]: My apologies. I meant to say `getALocalSource`, which is already exposed on `Node`.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (tausbn) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:140]: Judging by [this](https://lgtm.com/query/5057557791518890658/), all `ParameterNode`s are in fact `LocalSourceNode`s already.

**Comment by yoff** [comment]
The desirable api enhancements identified here are in #5398.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:80]:
```suggestion
When asking for local data flow or taint propagation between two expressions as above, you would normally constrain the expressions to be relevant to a certain investigation. The next section will give some concrete examples, but there is a more abstract concept that we should call out explicitly, namely that of a local source.

A local source is a data-flow node with no local data flow into it. As such, it is a local origin of data flow, a place where a new value is created. This includes parameters (which only receive global data flow) and most expressions (because they are not value-preserving). Restricting attention to such local sources gives a much lighter and more performant data-flow graph and in most cases also a more suitable abstraction for the investigation of interest. The class ``LocalSourceNode`` represents data-flow nodes that are also local sources. It comes with a useful member predicate ``flowsTo(DataFlow::Node node)``, which holds if there is local data flow from the local source to ``node``.
```

A few small edits

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:116]:
```suggestion
We see that we get several data-flow nodes for an expression as it flows towards a call (notice repeated locations in the ``call`` column). We are mostly interested in the "first" of these, what might be called the local source for the file name. To restrict attention to such local sources, and to simultaneously make the analysis more performant, we have the QL class ``LocalSourceNode``. We could demand that ``expr`` is such a node:
```

In general, we try to avoid words like "simply" in the docs. Just in case things aren't quite as simple as we hope 😉

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (shati-patel) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:277]:
```suggestion
For global flow, it is also useful to restrict sources to instances of ``LocalSourceNode``. The predefined sources generally do that.
```

**Review by shati-patel — COMMENTED** [review]
A few more small edits, but the docs look good overall 🥇

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:116]: Ah, this was in contrast to the following methods. Perhaps "directly" or "explicitly" is better. But no word also seems to work...

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:130]: Ah, done.

File: docs/codeql/codeql-language-guides/codeql-library-for-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/codeql-library-for-python.rst:24]: I think it would be good to mention DataFlow nodes and API-graph nodes here as well (since they are 2 new categories). If this is planned for an other PR, just let me know :wink: 

Adding two more categories would also take the count back to 4, so we don't need to fix the `divided into four` text on line 19 :joy:

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:23]: > all of which are subclasses of ``CfgNode`` which holds all those data-flow nodes which are associated with a control-flow node.

I'm not sure what the point you're trying to make with this text is. 

I think splitting it up to be it's own sentence, or moving it like in the suggestion could improve things a bit. But really, I'm unsure, since I don't quite understand the intent :sweat_smile: 

```suggestion
The local data flow library is in the module ``DataFlow``, which defines the class ``Node`` denoting any element that data can flow through. The ``Node`` class has a number of useful subclasses, such as ``ExprNode`` for expressions, ``CfgNode`` for control-flow nodes, ``CallCfgNode`` for function and method calls, and ``ParameterNode`` for parameters. You can map between data flow nodes and expressions/control-flow nodes using the member predicates ``asExpr`` and ``asCfgNode``:
```

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:286]: Should this section maybe come before the two examples? (that would feel more natural to me)

Oh, I see that C# does the same as we do right now... feel free to just leave it like this :+1:

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:292]:
```suggestion
   -  ``DataFlow::ExprNode`` - an expression behaving as a data flow node.
   -  ``DataFlow::CfgNode`` - a control-flow node behaving as a data flow node.
   -  ``DataFlow::CallCfgNode`` - a control-flow node for a function or method call behaving as a data flow node.
   -  ``DataFlow::ParameterNode`` - a parameter data flow node representing the value of a parameter at function entry.
```

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Looks really good :muscle: A few minor things, then it's ready to be merged from my point of view :1st_place_medal: 

Personally, I don't feel strongly about `flowsTo` vs `DataFlow::localFlow` (or the taint version of this). I suggest we don't postpone getting the PR merged because of this, and can solve this in a follow up discussion next week.

**Comment by yoff** [comment]
Sorry to keep updating this, but I was reminded by https://github.com/github/codeql-python-team/issues/319 that we needed to describe attributes and saw a natural place to do it. I think now that the remaining description of type trackers are best left to a separate PR, so this is the final tweak. 🤞

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:304]: I think it's great to mention `Concepts` :+1: 

Do you want to build an exhaustive list here? (meaning we need to update it every time we add a new concept) otherwise we _could_ give a few examples, and say that more can be found in the file :shrug:

_(I'm leaning towards it being a pain point to list them all manually, but interested to see how you feel about it)_

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:286]: I actually already changed this :-)

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:304]: I am unsure, actually. When I saw that the list was not very long, I just took them all, as I felt it could be useful to have such an index somewhere, and this is probably one of the first places people would look.

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:23]: The intent was to also express the type hierarchy, but I agree that it is trying to do too much. Subtly hinting at it, as in your suggestion, works much better.

File: docs/codeql/codeql-language-guides/codeql-library-for-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/codeql-library-for-python.rst:24]: Excellent point :-D I have added the two categories and given forward references (the reference to the API section commented out until #5389 is merged).

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:292]: Good suggestion, I have incorporated `CfgNode` and `CallCfgNode` into the hierarchy.

**Comment by calumgrant** [comment]
@xcorail would really like to open up Python to the bug bounty programme this week so could this be merged asap. Am I right in thinking that once this has been merged, the documentation is in a good enough state again?

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:308]: Continuing the discussion from https://github.com/github/codeql/pull/5357#discussion_r595090529, I think with the addition of this:

> The library ``Concepts`` (defined in module ``semmle.python.Concepts``) contain several subclasses of ``DataFlow::Node`` that are security relevant, such as ``FileSystemAccess`` and ``SqlExecution``.

we don't need to talk a great deal more about Concepts. Specifically, I think the list becomes too bloated with all those concepts, so I would recommend we don't include them. (I'm probably feeling 5/10 about this)

```suggestion
```

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (RasmusWL) [inline docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:299]:
```suggestion
    -  ``Attributes::AttrRead`` - an attribute read as a data flow node.
    -  ``Attributes::AttrWrite`` - an attribute write as a data flow node.
```

To keep things in line with the wording we've used for the other parts.

File: docs/codeql/codeql-language-guides/codeql-library-for-python.rst
Comment (RasmusWL) [reply docs/codeql/codeql-language-guides/codeql-library-for-python.rst:24]: Nice :+1:

**Review by RasmusWL — CHANGES_REQUESTED** [review]
juuuust a few last things :blush:

File: docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst
Comment (yoff) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:308]: So where do you propose this list be available? Or do you mean that no-one should maintain it and users should simply trawl `Concepts.qll` or rely on auto-complete?

Comment (RasmusWL) [reply docs/codeql/codeql-language-guides/analyzing-data-flow-in-python.rst:308]: discussed this in person, @yoff is feeling much stronger about this, so I'll bend to his will :+1:

**Review by RasmusWL — APPROVED** [review]

**Pull request merged**
